### PR TITLE
Bookmarks - API - initial stubs for the API tests

### PR DESCRIPTION
**Pull request by rplevka**
Initial API CRUD test stubs for Bookmark entity.

### CODE REVIEW

**Comment by sthirugn** [comment]
ACK

File: tests/foreman/api/test_bookmarks.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+from robottelo.test import APITestCase
+
+
+class BookmarkTestCase(APITestCase):
+    """Test for common Bookmark operations via API"""
+
+    # CREATE TESTS
+    @stubbed()
+    @tier1
+    def test_positive_create_bookmark(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: I don't think the name of the test needs to include the name of the entity being created. It should be implicit from the module name. Instead, based on the docstring, perhaps the name should be `test_positive_create_with_name_and _contoller`?

File: tests/foreman/api/test_bookmarks.py [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_positive_create_bookmark_public(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_positive_create_public`?

File: tests/foreman/api/test_bookmarks.py [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_create_bookmark_no_name(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_create_no_name`?

File: tests/foreman/api/test_bookmarks.py [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+        """Create a bookmark with empty name
+
+        @Feature: Scoped Search Bookmark Create
+
+        @Steps:
+
+        1. Create a bookmark with a providing no name
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/with a//

File: tests/foreman/api/test_bookmarks.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_create_bookmark_no_querye(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_create_no_query`?

File: tests/foreman/api/test_bookmarks.py [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+        """Create a bookmark with empty query
+
+        @Feature: Scoped Search Bookmark Create
+
+        @Steps:
+
+        1. Create a bookmark with a providing no query
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/a providing/providing an empty query

File: tests/foreman/api/test_bookmarks.py [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_create_bookmark_same_name(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_create_same_name`?

File: tests/foreman/api/test_bookmarks.py [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+        """Create bookmarks with the same names
+
+        @Feature: Scoped Search Bookmark Create
+
+        @Setup:
+
+        1. Create a bookmark of a random name
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a/with a

File: tests/foreman/api/test_bookmarks.py [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Steps:
+
+        1. Create a bookmark with providing the same name as the pre-created
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: 1. Create a new Bookmark using a random name
1. Create a second Bookmark, using the same name as the previous Bookmark. Assert that an error is raised
1. List Bookmarks. Assert that the Bookmark created is present and there's only one listed

File: tests/foreman/api/test_bookmarks.py [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_create_bookmark_public_null(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_create_ommit|null_public`?

File: tests/foreman/api/test_bookmarks.py [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+        """Create a bookmark with omitting the public parameter
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/with//

File: tests/foreman/api/test_bookmarks.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Feature: Scoped Search Bookmark Create
+
+        @Steps:
+
+        1. Create a bookmark of a random name, random query and omit the
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a/with a

File: tests/foreman/api/test_bookmarks.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    # UPDATE TESTS
+    @stubbed()
+    @tier1
+    def test_positive_update_bookmark_name(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_positive_update_name`?

File: tests/foreman/api/test_bookmarks.py [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+        """Update a bookmark
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create a bookmark of a random name with random query
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a random name with random query/with a random name and random query

File: tests/foreman/api/test_bookmarks.py [hunk 15]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Steps:
+
+        1. Update the name of the pre-created bookmark
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: 1. Create a new Bookmark with a random name
1. Update the previously created Bookmark with another random name

File: tests/foreman/api/test_bookmarks.py [hunk 16]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_update_bookmark_name(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_update_name`?

File: tests/foreman/api/test_bookmarks.py [hunk 17]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create 2 bookmarks of random names with random query
+
+        @Steps:
+
+        1. Update the name to one of the other pre-created bookmarks
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: 1. Try to update the name of the first (or second) Bookmark created in the Setup with the name of the second (or first) Bookmark

File: tests/foreman/api/test_bookmarks.py [hunk 18]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_update_bookmark_name_empty(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_update_empty_name`?

File: tests/foreman/api/test_bookmarks.py [hunk 19]
```diff
@@ -0,0 +1,290 @@
...
+        """Update a bookmark with an empty name
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create a bookmark of a random name with random query
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a random name with random query/with a random name and random query

File: tests/foreman/api/test_bookmarks.py [hunk 20]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Steps:
+
+        1. Update the name of the pre-created bookmark to an empty value
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: 1. Update the name of one of the previously created Bookmarks to an empty value

File: tests/foreman/api/test_bookmarks.py [hunk 21]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_positive_update_bookmark_query(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_positive_update_query`?

File: tests/foreman/api/test_bookmarks.py [hunk 22]
```diff
@@ -0,0 +1,290 @@
...
+        """Update a bookmark query
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create a bookmark of a random name with random query
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a random name with random query/with a random name and random query

File: tests/foreman/api/test_bookmarks.py [hunk 23]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_negative_update_bookmark_query_empty(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_negative_update_empty_query`?

File: tests/foreman/api/test_bookmarks.py [hunk 24]
```diff
@@ -0,0 +1,290 @@
...
+        """Update a bookmark with an empty query
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create a bookmark of a random name with random query
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a random name with random query/with a random name and random query

File: tests/foreman/api/test_bookmarks.py [hunk 25]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Steps:
+
+        1. Update the query of the pre-created bookmark
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: 1. Update the query of the pre-created bookmark with an empty value

File: tests/foreman/api/test_bookmarks.py [hunk 26]
```diff
@@ -0,0 +1,290 @@
...
+
+        @Status: Manual
+        """
+        for entity in BOOKMARK_ENTITIES:
+            with self.subTest(entity['controller']):
+                pass
+
+    @stubbed()
+    @tier1
+    def test_positive_update_bookmark_public(self):
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: `test_positive_update_public`?

File: tests/foreman/api/test_bookmarks.py [hunk 27]
```diff
@@ -0,0 +1,290 @@
...
+        """Update a bookmark public state
+
+        @Feature: Scoped Search Bookmark Update
+
+        @Setup:
+
+        1. Create 2 bookmarks of a random names with random queries, one public
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: s/of a random name with random query/with a random name and random query

File: tests/foreman/api/test_bookmarks.py [hunk 28]
```diff
@@ -0,0 +1,290 @@
...
+        and one private
+        2. Create a non-admin user with 'viewer' role
+
+        @Steps:
+
+        1. List the bookmarks using the pre-created user's credentials
```
Comment (omaciel) [inline tests/foreman/api/test_bookmarks.py]: Hmmm just to see if I understood this:

1. Using the `admin` user, list the Bookmarks previously created. Assert that their `public` fields match the values used during their creation.
1. Update the Bookmarks previously created, setting their `public` values to the opposite value of their original values (ie. `True > False` and `False > True`)
1. Using the `non-admin` user, list the Bookmarks. Assert that their `public` fields match the new values used during the `update` steps

Comment (rplevka) [inline tests/foreman/api/test_bookmarks.py]: @omaciel after inspecting the `private/public` feature in more detail, It makes no sense to create an additional user. The `private/public` feature does not have an effect on the common bookmark listing, it only applies to visibility at the entity page which is not applicable for the `API`.
this test should be replaced by 2 additional tests:
`test_positive_update_public`
`test_positive_update_private`
where only updating the state would be tested and asserted.

**Comment by rplevka** [comment]
All comments addressed

**Comment by omaciel** [comment]
@rplevka I think you missed one small docstring change, but no biggie. ACK

**Pull request merged**